Add VASP recipe for GH200

AdhocMan opened the pull request:
Initial recipe for VASP on the A100 and GH200 architecture.
The VASP source code and binaries are restricted to licensed users only. 
JFrog has been configured such that only users in the `vasp6` group and the CI pipeline should have access rights.

RMeli commented:
I think this needs a documentation page too (`docs/uenv-vasp`). In particular, it needs to have the "Licensing Terms and Conditions" that we have in the [user portal](https://user.cscs.ch/computing/applications/vasp/). In https://github.com/eth-cscs/alps-uenv/pull/85 I added it as a `danger` box.

AdhocMan commented:
> I think this needs a documentation page too (`docs/uenv-vasp`). In particular, it needs to have the "Licensing Terms and Conditions" that we have in the [user portal](https://user.cscs.ch/computing/applications/vasp/). In #85 I added it as a `danger` box.

Thanks for the suggestion, I've added some documentation about licensing.

The A100 recipe has now been removed for now, since it seems to be a while until we can test it again. The only difference is the cuda architecture spec, so it will be easy to add again.

AdhocMan commented:
cscs-ci run alps;system=santis;uarch=gh200;uenv=vasp:v6.4.2

AdhocMan commented:
cscs-ci run alps;system=santis;uarch=gh200;uenv=vasp:v6.4.2

AdhocMan commented:
cscs-ci run alps;system=santis;uarch=gh200;uenv=vasp:v6.4.2

AdhocMan commented:
cscs-ci run alps;system=santis;uarch=gh200;uenv=vasp:v6.4.2

AdhocMan commented:
cscs-ci run alps;system=santis;uarch=gh200;uenv=vasp:v6.4.2

AdhocMan commented on recipes/vasp/v6.4.2/gh200/compilers.yaml:
```diff
@@ -0,0 +1,9 @@
+  spec: gcc@12.2
+gcc:
+  specs:
+  - gcc@12.2
```
The nvhpc, gcc and cuda versions are all tied together. The nvhpc version I tested with comes with bundled cuda 12.2, which only supports up to gcc 12.2 as well. Unfortunately, spack doesn't recognize this, so I'm using a matching cuda spec to make sure MPI is built with the same cuda version.
The spack architecture flags for all gcc 12 versions are the same and only gcc 13 has neoverse v2 specific optimizations. Technically, a neoverse v2 specific flag was added in gcc 12.3.0, but the official gcc release notes only mention support with gcc 13. They are also missing from the documentation (https://gcc.gnu.org/onlinedocs/gcc-12.3.0/gcc/ARM-Options.html). That's why the same Arm architecture flags for all gcc 12 versions are used. These should cover the instruction set Grace is capable of, but don't quite have perfect tuning parameters.
In the case of this vasp recipe, all performance critical packages like vasp itself and Openblas are build with nvhpc anyway, which includes neoverse v2 optimizations.

bcumming replied:
I will start by saying that the configuration is good for a first deployment - we can go with these software versions, and release an update with different cuda/gcc/nvhpc versions later if we need to.

We also need some documentation with recommended versions of compilers, runtimes, etc that we recommend all uenv use if possible. That is on me!

NVIDIA gave us feedback that 12.3 was the first version of GCC  that should be used with Grace (even if it isn't explicit support, it might well tested).

Have you tried nvhpc 24.1 or 24.3? They are both supported on Santis with Stackinator and Spack v0.21.

When a cuda spec, e.g. `cuda@12.3`, then that version will be installed and used, instead of the version inside the `nvhpc` SDK (at least it is in other projects that we use `nvhpc` to build the Fortran OpenACC code, e.g. ICON).

The most recent software versions that we haver tested with success (not guarenteed to work for all use cases) is:
```
nvhpc@24.1 / 24.3
gcc@13.2
cuda@12.4
```

AdhocMan replied:
I think the advice from Nvidia with regards to GCC 12.3 only makes sense if you use `-mpcu=neoverse-v2`, which we don't when using spack (since it's an undocumented feature, so spack doesn't use it).

Using GCC 13 would be ideal, but it requires cuda 12.4, as you suggested. But nvhpc 24.3 only comes with cuda 12.3 (https://docs.nvidia.com/hpc-sdk/archive/24.3/hpc-sdk-release-notes/index.html). The way vasp is currently build, one has to use a bundled cuda version. And when I tested with stackinator, spack does not appear to find the bundled cuda version (at least when nvhpc is installed as a @gcc package and used as a compiler). I ended up removing the nvhpc dependency from the vasp package, since we would then end up with two nvhpc packages in the image. So an explicit cuda package version is used instead for the MPI package. I'd be curious to learn how other OpenACC codes solved this.

Going forward, I'd like to upgrade the compiler / cuda version. I could test nvhpc 24.3 and cuda 12.3 with gcc 12.2. It will take about two days to run the verification tests however.

AdhocMan commented:
cscs-ci run alps;system=santis;uarch=gh200;uenv=vasp:v6.4.2

AdhocMan commented on recipes/vasp/v6.4.2/gh200/environments.yaml:
```diff
@@ -0,0 +1,59 @@
+      - python%gcc ~bz2 ~dbm ~lzma ~nis ~pyexpat ~sqlite3 ~ssl ~uuid ~zlib
+  views:
+    default:
+    develop:
```
Licensed users have access to the source code and VASP has a lot of build options. I think it's quite possible, that someone would like to compile their own version.

sekelle replied:
First user has already requested this.

AdhocMan commented on docs/uenv-vasp.md:
```diff
@@ -0,0 +1,7 @@
+# VASP
+
+VASP (Vienna Ab initio Simulation Package) is a software package for performing ab initio quantum mechanical calculations.
```
I think it's still not quite clear how we will make these images available (jfrog, local filesystem or something else?). I'd like to wait with documenting this until we know more about how the filesystem on the new machine will look like and if placing the images in some directly accessible storage location is possible.

AdhocMan replied:
I've added some documentation on how to use it

AdhocMan commented:
cscs-ci run alps;system=santis;uarch=gh200;uenv=vasp:v6.4.2

AdhocMan commented:
I've tested upgrading compilers and discovered some issues:
- OpenBLAS does not seem to compile with NVHPC 23.9 and later using spack. One of the tests always segfaults. Also happens with the latest version of OpenBLAS
- The recipe does not compile with NVHPC 24.1 and 24.3. An error in an OpenACC section of VASP occurs.

So I've updated the recipe to NVHPC 23.9 and NVPL libraries for BLAS and LAPACK.

RMeli commented on docs/uenv-vasp.md:
```diff
@@ -0,0 +1,34 @@
+
+## Accessing VASP images
+
+TODO: Describe access to images. Not yet finalized.
```
```suggestion
!!! failure

    Describe access to images. Not yet finalized.
```
to make this a bit more prominent, so that we do not forget.

AdhocMan commented:
cscs-ci run alps;system=santis;uarch=gh200;uenv=vasp:v6.4.2

AdhocMan commented:
Some of the tests in the included testsuite of VASP fail when compiled using the architecture flag `-tp neoverse-v2`. Using `-tp host` instead fixed the issues.
So the configuration has been changed to use a modified spack based on v0.21.0, where the architecture flag for NVHPC is is changed.

bcumming commented:
cscs-ci run alps;system=todi;uarch=gh200;uenv=vasp:v6.4.2

AdhocMan commented:
cscs-ci run alps;system=todi;uarch=gh200;uenv=vasp:v6.4.2
